core/src/Streamly/Internal/Data/Serialize/TH.hs

```diff
@@ -381,44 +380,47 @@ mkSerializeDec (Config {..}) headTy cons = do
 --       ])
 -- @
 deriveSerializeInternal ::
-       Config -> Cxt -> Type -> [DataCon] -> Q [Dec]
-deriveSerializeInternal conf preds headTy cons = do
+       Config -> Type -> [DataCon] -> ([Dec] -> Q [Dec]) -> Q [Dec]
+deriveSerializeInternal conf headTy cons next = do
     sizeDec <- mkSizeDec conf headTy cons
     peekDec <- mkDeserializeDec conf headTy cons
     pokeDec <- mkSerializeDec conf headTy cons
     let methods = concat [sizeDec, peekDec, pokeDec]
-    return [plainInstanceD preds (AppT (ConT ''Serialize) headTy) methods]
+    next methods
 
--- | Similar to 'deriveSerialize,' but take a 'Config' to control how
--- the instance is generated.
+-- | Similar to 'deriveSerialize' but take a 'Config' to control how the
+-- instance is generated.
 --
--- Usage: @$(deriveSerializeWith config ''CustomDataType)@
-deriveSerializeWith :: Config -> Name -> Q [Dec]
-deriveSerializeWith conf@(Config {..}) name = do
-    dt <- reifyDataType name
-    let preds = map (unboxPred . VarT) (filterOutVars (dtTvs dt))
-        headTy = appsT (ConT name) (map substituteVar (dtTvs dt))
-        cons = dtCons dt
-    deriveSerializeInternal conf preds headTy cons
+-- Usage:
+-- @
+-- $(deriveSerializeWith
+--       config
+--       [d|instance Serialize a => Serialize (Maybe a)|])
+-- @
+deriveSerializeWith :: Config -> Q [Dec] -> Q [Dec]
+deriveSerializeWith conf mDecs = do
+    dec <- mDecs
+    case dec of
+        [inst@(InstanceD _ _ headTy [])] -> do
+            dt <- reifyDataType (getMainTypeName headTy)
+            let cons = dtCons dt
+            deriveSerializeInternal conf headTy cons (next inst)
+        _ -> errorUnsupported
+  where
+
+    next (InstanceD mo preds headTy []) methods =
+        pure [InstanceD mo preds headTy methods]
+    next _ _ = errorUnsupported
+
+    getMainTypeName = go . unwrap
+
+    unwrap (AppT (ConT _) r) = r
+    unwrap _ = error "getMainTypeName.unwrap: Undefined instance declaration."
```

Comment:
Should use the prefix `deriveSerializeWith` as that is what the user knows about. Also, does the message convey the specific error information that the user can understand, act upon and correct.

Reply:
Yes, we should be more specific.